text/0000-wallet-replay-protection.md

```diff
@@ -0,0 +1,319 @@
+
+#### Possible protected output generators implementations
+
+As we mentioned above, generator of `Protected` outputs would need to have `create` and `check` functions defined.
+
+There are a few options how to label outputs as `Protected` while still being able to identify them across difference devices:
+1. Call to `create` uses a new derivation path `P` that is used only for creating `Protected` outputs. Similarly `check` uses the same `P` to check whether an output is protected. Perhaps we could have `N` labels possible which would be labeled by the `r % N` result
+2. Call to `create` uses additional output information in the bytes that are available in the Bulletproofs to tell whether the output is protected. These bytes are right now 'zero' bytes meaning their binary representation is all zeros. We label an output as `Protected` by setting the `label` bit that is unused now to `1`. `Unprotected` outputs will have the label bit set to `0` which includes all the old outputs as well. The label bit can be read only by the owner of the output because the message that is encoded in the Bulletproof is already private. If we went such path, we would need to think of the possible drawbacks.
+
+In all cases, the output label should only be visible to the owner of the output. It seems necessary to have labeling information about the UTXO on the UTXO itself if we want it to be consistent with different wallet reusing the same seed. If the information is held only on the wallet side, then we hit much bigger issues because there comes a need for either a manual intervention and labeling or a robust solution for synchronization between wallets - which does not appear simple to build.
+
+### Simple bootstrapping of protected outputs
+
+We start off with all of our outputs marked as `Unprotected`. To create a `Protected` output out of nothing we create the following outputs:
+1. an `anchor` output that has a form `0*H + r*G` - generated from key derivation path `A`
+2. a set of protected outputs `PS` that hold our coins - generated from `GenP`
+
+We make a bootstrap transaction with `anchor` and `PS` outputs. The outputs in `PS` are labeled as `Protected` because they are joined in the same transaction output set as our anchor output. This transaction cannot be replayed because the anchor output will never be spent which means that `PS` outputs are safe from being recreated.
+
+_Note: Grin has a rule that an output that already exists in the UTXO set cannot be created - a transaction that would attempt to do that is invalid. This is why transactions that contain an anchor output can't be replayed._
+
+### Wallet transaction rules
+
+There are two known ways to protect against replay attacks using wallet rules. One is to utilize the `Protected` outputs in transactions, the other one is to keep a history of outputs that were spent and label as 'unsafe' those that the wallet does not remember creating.

```

Comment:
This seems to be the wrong place to discuss the alternative of the wallet remembering what it spent before.